### PR TITLE
Incremental yapf for precommit

### PR DESCRIPTION
**Pull request by ppolewicz**

### CODE REVIEW

File: pre-commit.sh [hunk 1]
```diff
@@ -1,5 +1,39 @@
-#!/bin/bash -eux
-yapf --in-place --recursive .
+#!/bin/bash -ux
+
+base_remote="$(1:-origin)"
```
Comment (konefalg) [inline pre-commit.sh]: Fragment of `pre-commit.sh` output:

```
++ 1:-origin
./pre-commit.sh: line 3: 1:-origin: command not found
+ base_remote=
```

Is it ok? The same problem occured for next two lines in the script. Variables `base_remote`, `base_branch`, `base_remote_branch` are not initialized.

File: pre-commit.sh [hunk 2]
```diff
@@ -1,5 +1,39 @@
...
+base_branch="$(1:-master)"
+base_remote_branch="$(1:-master)"
+
+set +e
+
+if [ "$(<.git/refs/heads/${base_branch})" != "$(<.git/refs/remotes/${base_remote}/${base_remote_branch})" ]; then
```
Comment (konefalg) [inline pre-commit.sh:9]: when variables `base_branch`, `base_remote` and `base_remote_branch` are uninitialized, it always returns false

File: pre-commit.sh [hunk 3]
```diff
@@ -1,5 +1,39 @@
...
+    base_commit=`diff -u "$head" "$master" | grep '^ ' | head -n 1 | cut -c 2-`  # the commit from which the master and topic (current) branch have diverged
+
+    # in below case, it would be E
+    #
+    #        A---B---C---I---K---L topic
+    #       /       /       /
+    #  D---E---F---G-------H---J master
+    #
+
+    git diff --name-only "${base_commit}..HEAD" #| grep '\.py$' | xargs yapf --in-place setup.py
```
Comment (konefalg) [inline pre-commit.sh]: Please uncomment invoking yapf

Comment (konefalg) [inline pre-commit.sh]: Please uncomment invoking yapf and remove setup.py from the end.

Comment (konefalg) [inline pre-commit.sh]: It will not work with uncommited changes. Maybe `git ls-files -m` will be helpful to fix this problem?

File: pre-commit.sh (hunk 1 above)
Comment (ppolewicz) [inline pre-commit.sh]: It should be { instead of (, I'll fix it on Tuesday

File: pre-commit.sh (hunk 2 above)
Comment (ppolewicz) [inline pre-commit.sh:9]: They will have default values

File: pre-commit.sh [hunk 4]
```diff
@@ -1,5 +1,39 @@
-#!/bin/bash -eux
-yapf --in-place --recursive .
+#!/bin/bash -ux
+
+base_remote="$(1:-origin)"
+base_branch="$(1:-master)"
```
Comment (ppolewicz) [inline pre-commit.sh]: This should be 2, not 1

File: pre-commit.sh [hunk 5]
```diff
@@ -1,5 +1,39 @@
...
+base_remote_branch="$(1:-master)"
```
Comment (ppolewicz) [inline pre-commit.sh]: This should be 3, not 1

File: pre-commit.sh (hunk 3 above)
Comment (ppolewicz) [inline pre-commit.sh]: If you want support for uncommitted changes, you can add that yourself

Comment (ppolewicz) [inline pre-commit.sh]: setup.py must stay so that commits with changes to non-python files don't cause yapf to be run without parameters

Comment (konefalg) [inline pre-commit.sh]: But in your code, if I understand correctly, you filtered out non-python files (`grep '\.py$'`), so that's unnecessary.

Comment (ppolewicz) [inline pre-commit.sh]: Check what happens when the list passed to xargs is empty

Comment (konefalg) [inline pre-commit.sh]: Please note, that not supporting for uncommited changes is a change of `pre-commit.sh` semantics. Contributors, who got used to using the script before those changes, will be confused when their code does not pass travis's tests after running pre-commit.

Comment (konefalg) [inline pre-commit.sh]: I see what happens when list passed to xargs is empty, but this is only corner case and another solution should be implemented to avoid starting yapf on the whole project. Do you think that running yapf on file `setup.py` many times is right?

Comment (ppolewicz) [inline pre-commit.sh]: As it was low effort and it sped up yapf by orders or magnitude, I thought it is ok. I have now changed it so yapf is only run when necessary.
